hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -881,4 +894,113 @@ private static RetryTemplate createRetryTemplate() {
+            invalidateDistributionSetsInTransaction(distributionSetInvalidation, tenant);
+        }
+    }
+
+    private void invalidateDistributionSetsInTransaction(final DistributionSetInvalidation distributionSetInvalidation,
+            final String tenant) {
+        DeploymentHelper.runInNewTransaction(txManager, tenant + "-invalidateDS", status -> {
+            distributionSetInvalidation.getSetIds().forEach(setId -> invalidateDistributionSet(setId,
+                    distributionSetInvalidation.getCancelationType(), distributionSetInvalidation.isCancelRollouts()));
+            return 0;
+        });
+    }
+
+    private void invalidateDistributionSet(final long setId, final CancelationType cancelationType,
+            final boolean cancelRollouts) {
+        final JpaDistributionSet set = (JpaDistributionSet) distributionSetManagement.getValidAndComplete(setId);
+        set.invalidate();
+        distributionSetRepository.save(set);
+        LOG.debug("Distribution set {} set to invalid", set.getId());
+
+        if (cancelationType != CancelationType.NONE || cancelRollouts) {
+            cancelRollouts(set);
+        }
+
+        if (cancelationType != CancelationType.NONE) {
+            cancelActions(cancelationType, set);
+        }
+
+        cancelAutoAssignments(setId);
+    }
+
+    private void cancelRollouts(final DistributionSet set) {
+
+        // stop all rollouts for this distribution set
+        rolloutRepository.findByDistributionSetAndStatusIn(set, ROLLOUT_STATUS_STOPPABLE).forEach(rollout -> {
+            final JpaRollout jpaRollout = (JpaRollout) rollout;
+            jpaRollout.setStatus(RolloutStatus.STOPPING);
+            rolloutRepository.save(jpaRollout);
+            LOG.debug("Rollout {} stopped", jpaRollout.getId());
+        });
+
+    }
+
+    private void cancelActions(final CancelationType cancelationType, final DistributionSet set) {
+        actionRepository.findByDistributionSetAndActiveIsTrue(set).forEach(action -> {
+            final JpaAction jpaAction = (JpaAction) action;
+            cancelAction(jpaAction.getId());
+            LOG.debug("Action {} canceled", jpaAction.getId());
+        });
+        if (cancelationType == CancelationType.FORCE) {
+            actionRepository.findByDistributionSetAndActiveIsTrue(set).forEach(action -> {
+                final JpaAction jpaAction = (JpaAction) action;
+                forceQuitAction(jpaAction.getId());
+                LOG.debug("Action {} force canceled", jpaAction.getId());
+            });
+        }
+    }
+
+    private void cancelAutoAssignments(final long setId) {
+        // remove distribution set from all auto assignments
+        targetFilterQueryRepository.unsetAutoAssignDistributionSetAndActionType(setId);
+        LOG.debug("Auto assignments for distribution sets {} deactivated", setId);
+    }
+
+    @Override
+    public long countRolloutsForInvalidation(final Collection<Long> setIds) {
```

Comment:
To find out the side effectes of the invalidate call a caller has to know that he has to call up to 3 methods dependend on the input parameters of the invalidate call. So the caller needs some knowledge how the invalidation internally works. To hide this information from the caller the invalidate method could have an additional parameter "dryRun" and return a container object with the number of all affected entities.

Reply:
This will be changed with an upcoming PR